### ISSUE TITLE
Memory limit is not honoured in some situations

### ISSUE THREAD

**Issue by veluca93** [cms-dev/cms#685]
![screenshot from 2016-12-15 15-52-42](https://cloud.githubusercontent.com/assets/1896483/21228617/bef03212-c2de-11e6-8c5f-769877c524ce.png)
As you can see in this screenshot, this submission used around 700 MiB of main memory, but the memory limit for the task is 256 MiB.
I suspect this issue arises from submissions that, in some way or the other, behave like
```while (true) putc('a');```
but this seems strange to me, as I am quite sure I checked this at some point.

**Comment by andreyv** [comment]
Maybe this is related to `IsolateSandbox` passing the `--mem` parameter to `isolate` instead of `--cg-mem`. The situation you describe looks like the program filling the kernel filesystem cache, which is not a part of the program's address space. Observe how memory usage in `IsolateSandbox` is obtained from `cg-mem`, which would also include the filesystem cache pages.

In this case, it is not clear what a good value of `--cg-mem` would be, or whether control group memory needs to be limited at all (after all, caches don't hog memory permanently and allow for faster I/O).

**Comment by veluca93** [comment]
That would make sense, but notice that this did result in a file of ~700mb
being created - a clever and determined cheater could use that to bypass
the memory limit (especially in tasks with long time limits)

**Comment by andreyv** [comment]
Good point. There is an `--fsize` parameter to `isolate`, but it only limits the size of one file.

Right now I can think of these alternatives for improvement:
- Use filesystem quotas/subvolumes (probably not a good way).
- Put the sandbox in a `tmpfs`, where capacity can be easily limited.
- Disallow any write access to the filesystem and use I/O through standard input/output (best way).

But this is a topic for a separate issue.

**Comment by veluca93** [comment]
On a related note (but that maybe needs its own issue), if the folder where sandboxes are becomes full, cmsWorker won't delete the sandbox even if keep_sandbox is false - these two problems combined can result in a DoS-type issue (in fact, that was how I discovered this issue in the first place)

**Comment by andreyv** [comment]
For the record, #693 changes `--mem` to `--cg-mem`.

**Comment by asdacap** [comment]
Have you tried enabling `swapaccount=1` kernel config? Its not enabled by default, at least on ubuntu 14.04. It seems that --cg-mem will only limit the actual physical RAM usage if the config is not enabled, which means the program can use the swap and exceed the specified memory limit.

**Comment by stefano-maggiolo** [comment]
Also make sure you have no swap in the machines running the workers

**Comment by lerks** [comment]
Has this issue ever occurred again on a swapless system after #693 was applied?

**Comment by veluca93** [comment]
On a swapless system (or with #693, not sure which is what) it works as expected.
